lib/wavefront/client/common/proxy_connection_handler.rb

```diff
@@ -13,55 +14,101 @@ class ProxyConnectionHandler
     # @param port [Integer] Proxy Port
-    def initialize(address, port)
-      @failures = AtomicInteger.new
-      @address = address
-      @port = port
-      @reconnecting_socket = nil
+    def initialize(address, port, internal_store, retries: 3, retry_interval: 0.5)
+      @internal_store = internal_store
       @lock = Mutex.new
+
+      @client_options = {
+        server: "#{address}:#{port}",
+        # disable client retries as we collect failure counts
+        connect_retry_count: 0,
+        retry_count: 0
+        # use default timeouts
+        #   connect: 10s; write: 60s
+      }.freeze
+      @retries = retries
+      @retry_interval = retry_interval
+
+      @errors = @internal_store.counter('errors')
+      @connect_errors = @internal_store.counter('connect.errors')
     end
 
     # Open a socket connection to the given address:port
     def connect
       @lock.synchronize do
-        # Create a new socket connection if not already done
-        @reconnecting_socket = TCPSocket.open(@address, @port) if @reconnecting_socket.nil?
+        last_error = nil
+        @retries.times do |_ri|
+          begin
+            connect_locked
+          rescue Net::TCPClient::ConnectionFailure => e
+            Wavefront.logger.warn "Connection failure: #{e.cause}"
+            @connect_errors.inc
+            last_error = e.cause
+          rescue Net::TCPClient::ConnectionTimeout, Net::TCPClient::WriteTimeout => e
+            Wavefront.logger.warn "Warning: #{e.cause}"
+            @connect_errors.inc
+            last_error = e.cause
+          rescue StandardError => e
+            Wavefront.logger.warn "Unexpected Error: #{e}\n\t#{e.backtrace.join("\n\t")}"
+            last_error = e
+          ensure
+            sleep @retry_interval
+          end
+        end
+        @errors.inc unless last_error.nil?
+        raise SendError, last_error unless last_error.nil?
       end
     end
 
     # Close socket if it's open now.
     def close
-      @reconnecting_socket.close if @reconnecting_socket
+      @client&.close
     end
 
-    # Returns number of failures
     def failure_count
-      @failures.value
-    end
-
-    # Increment failure count
-    def increment_failure_count
-      @failures.increment
+      @errors.value
     end
 
     # Send data via proxy.
-
     # @param line_data [String] Data to be sent
     # @param reconnect [Boolean] If it's the second time trying to send data
-    def send_data(line_data, reconnect=true)
-      begin
-        connect unless @reconnecting_socket
-        @reconnecting_socket.puts(line_data.encode('utf-8'))
-      rescue SocketError => error
-        if reconnect
-          @reconnecting_socket = nil
-          # Try to resend
-          send_data(line_data, false)
-        else
-          # Second time trying failed
-          raise error
+    def send_data(line_data)
+      # tcp writes may not be thread safe.
+      @lock.synchronize do
```

Comment:
I have a concern with this line of code. We cannot synchronize here as this has to be high throughput.
See how Python does it - https://github.com/wavefrontHQ/wavefront-sdk-python/blob/56fb0da27fc15d2fa398dd268611569a92fe5196/wavefront_sdk/common/proxy_connection_handler.py